File: Java/Problem_070.java
import java.util.*;

public class Euler70 {
    static int[] sieve(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;

        for (int i = 2; (long)i * i <= limit; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= limit; j += i) isPrime[j] = false;
            }
        }

        int count = 0;
        for (int i = 2; i <= limit; i++) if (isPrime[i]) count++;
        int[] primes = new int[count];
        int idx = 0;
        for (int i = 2; i <= limit; i++) if (isPrime[i]) primes[idx++] = i;
        return primes;
    }

    static int[] sig(int x) {
        int[] c = new int[10];
        while (x > 0) {
            c[x % 10]++;
            x /= 10;
        }
        return c;
    }

    static boolean sameSig(int a, int b) {
        return Arrays.equals(sig(a), sig(b));
    }

    public static void main(String[] args) {
        final int LIMIT = 10_000_000;
        int[] primes = sieve(LIMIT);

        long bestN = 0;
        long bestNum = 1;
        long bestDen = 0;

        for (int i = 0; i < primes.length; i++) {
            long p = primes[i];
            if (p * p >= LIMIT) break;

            for (int j = i; j < primes.length; j++) {
                long q = primes[j];
                long n = p * q;
                if (n >= LIMIT) break;

                long phi = (p - 1) * (q - 1);

                if (sameSig((int)n, (int)phi)) {
                    if (bestDen == 0 || n * bestDen < bestNum * phi) {
                        bestN = n;
                        bestNum = n;
                        bestDen = phi;
                    }
                }
            }
        }

        System.out.println(bestN); // 8319823
    }
}

File: CPP/Problem_070.cpp
#include <iostream>
#include <vector>
#include <array>
using namespace std;

static vector<int> sieve(int limit) {
    vector<char> isPrime(limit + 1, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; 1LL*i*i <= limit; i++) {
        if (isPrime[i]) {
            for (int j = i*i; j <= limit; j += i) isPrime[j] = false;
        }
    }
    vector<int> primes;
    primes.reserve(limit / 10);
    for (int i = 2; i <= limit; i++) if (isPrime[i]) primes.push_back(i);
    return primes;
}

static array<int,10> sig(int x) {
    array<int,10> c{};
    c.fill(0);
    while (x) {
        c[x % 10]++;
        x /= 10;
    }
    return c;
}

int main() {
    const int LIMIT = 10'000'000;
    vector<int> primes = sieve(LIMIT);

    long long bestN = 0;
    long long bestNum = 1;
    long long bestDen = 0;

    for (size_t i = 0; i < primes.size(); i++) {
        long long p = primes[i];
        if (p * p >= LIMIT) break;

        for (size_t j = i; j < primes.size(); j++) {
            long long q = primes[j];
            long long n = p * q;
            if (n >= LIMIT) break;

            long long phi = (p - 1) * (q - 1);

            if (sig((int)n) == sig((int)phi)) {
                if (bestDen == 0 || n * bestDen < bestNum * phi) {
                    bestN = n;
                    bestNum = n;
                    bestDen = phi;
                }
            }
        }
    }

    cout << bestN << "\n"; // 8319823
    return 0;
}
